### DISCUSSION TITLE
ngspice 42 supports XSPICE Digital Logic

### DISCUSSION THREAD

**Discussion by tomhajjar** [ra3xdh/qucs_s#635] [General]
"With ngspice-42 the simulation of 74xxx is established. PSPICE or MicroCap U devices will be translated into XSPICE event-based simulation."
https://sourceforge.net/p/ngspice/discussion/120973/thread/e6442c59ce/?limit=25&page=1#ebba

Do the models in Qucs-S support using ngspice/XSPICE as stated in #97? Is there an example using XSPICE logic?

I have a very old project using mixed analog/digital I want to try. It's a binary trainline driver for locomotives.

![SSR](https://github.com/ra3xdh/qucs_s/assets/102838604/b3414cf5-d5df-4993-9c69-24258c77a277)

**Comment by ra3xdh** [comment]
Qucs-S currently provides the symbols only for ideal logic gates. Some of these devices (logic gates itself and triggers) could be used in analog mode. The Ngspice 74XX devices library will work if it will be wrapped in Qucs-S library. You have to draw symbol and attach SPICE model for every device. It is not a complex task, but requires a lot of work.

**Comment by ra3xdh** [comment]
Also LTSpice has more advanced digital library for analog mode: https://github.com/texane/power_inverter/blob/master/ltspice/logic/74hc/74hc.lib It includes input/output transistors model. The MicroCAP version that is recommended for Ngspice includes only ideal gates with delay. The LTspice library also may work, but it must be ported manually because of incompatible logic gates definition syntax.

**Comment by tomhajjar** [comment]
First cut at logic gates for mixed digital-analog simulations. I found a 74HC library modified for ngspice, "74HCng_short_2.lib". The mods are extensive using adc_bridge and dac_bridge. I haven't gotten the OR/NOR gates to work. 

The library is discussed here:
https://sourceforge.net/p/ngspice/discussion/120973/thread/e6442c59ce/?limit=25#798e/ac37

Holger claims ngspice 42 allows PSpice and MicroCap U devices to work. I don't have those libraries to test. 

https://ngspice.sourceforge.io/modelparams.html 

![2024-03-11_215835](https://github.com/ra3xdh/qucs_s/assets/102838604/59115f9c-be8d-47b5-8dbc-cd4a25cee607)
![2024-03-11_220743](https://github.com/ra3xdh/qucs_s/assets/102838604/c3c10f4d-6167-4161-b206-00365912db5c)
![2024-03-11_220755](https://github.com/ra3xdh/qucs_s/assets/102838604/05d6cbd4-e6d1-4c3f-9fe6-25bdf4ae00b2)

**Comment by ra3xdh** [comment]
> I haven't gotten the OR/NOR gates to work.

Adding output resistance between gates may help. The XSPICE models has source output and may produce infinite values when connected to other logic input. 

> Holger claims ngspice 42 allows PSpice and MicroCap U devices to work. I don't have those libraries to test.

Did you tried the libraries from here: https://ngspice.sourceforge.io/model-parameters/MicroCap-LIBRARY.7z The 74HC series is inside dig000.lib file in the archive.

**Comment by tomhajjar** [comment]
I found my errors in the OR/NOR test circuits and uploaded the corrected files above. I had brain fog...

I found the MicroCap libraries in the link below. It would be nice to see an example using them with ngspice. 
https://gotroot.ca/spectrum/www.spectrum-soft.com/download/download.html

I would like to get the symbols for all the logic devices. Is there a way to get them in the "sch" format so I can make libraries? Qucs makes creating curved lines from the Elliptic Arc very tedious since one cannot control the end points. I end up cutting sections from inductor symbol files.

**Comment by tomhajjar** [comment]
I followed Holger's advice for using the MicroCap library "dig000.lib". I am having lot's of issues and some is due to Qucs-S and some may be due to ngspice. 

To start I created "NAND_hc.lib" from "dig000.lib". It is just 74HC00 and 74HC132. I made the sub "NAND_hc_mc.sch" using NAND_hc.lib. It works in "NAND_hc_mc_tran.sch". 

I then made a new sub "74HC00_mc.sch" using "dig000.lib". It doesn't work in "74HC00_mc_tran.sch" and generates a lot of errors.

I then made "74HC00_mc_tran_mod.sch" using "NAND_hc_mc.sch" and it doesn't work but generates a one error.

I am baffled how "NAND_hc_mc.sch" works in "NAND_hc_mc_tran.sch" but not in "74HC00_mc_tran_mod.sch".

One clue is set "ngbehavior=psa" is not being included in 2 out of 3 netlists yet the errors are very different.

Holger told me what to do in https://sourceforge.net/p/ngspice/discussion/120973/thread/e6442c59ce/?page=1
![2024-03-12_132600](https://github.com/ra3xdh/qucs_s/assets/102838604/d5001155-908e-45c5-bd04-e385561737b6)
![2024-03-12_132238](https://github.com/ra3xdh/qucs_s/assets/102838604/1010d936-5969-48a1-8fc8-0843333cbf2a)
![2024-03-12_132311](https://github.com/ra3xdh/qucs_s/assets/102838604/0f0add54-b163-4c92-b2f8-1893a75f3923)
![2024-03-12_132347](https://github.com/ra3xdh/qucs_s/assets/102838604/002286ce-46a8-4858-99c9-4c2e86bded89)

**Comment by ra3xdh** [comment]
> some is due to Qucs-S and some

Use SpiceFile instead of SpiceLib comp and you will get rid of `.SUBCKT nesting` warning.

**Reply by tomhajjar** [reply]
I don't know what one is "SpiceFile" and how is this causing the errors like no "ngbehavior=psa"?

![2024-03-12_135711](https://github.com/ra3xdh/qucs_s/assets/102838604/d0520f7f-817e-4848-b57d-6083a75f4fae)

**Reply by ra3xdh** [reply]
> I don't know what one is "SpiceFile"

I meant `Spice netlist`

**Reply by tomhajjar** [reply]
If I use Spice netlist, it grabs the first device in the list. How does this work for a library full of devices?

**Reply by ra3xdh** [reply]
Copy paste a model form the library into a text file and attach to individual devices.

**Reply by tomhajjar** [reply]
Yikes. For a digital library that is a lot of work. You think this is the issue?

I made two netlists and only one works. The ONLY difference is the position of the line"

"XY1 in _net0 out NAND_hc_mc"

Is it possible we have two issues, Qucs-S is not writing the .spiceinit "ngbehavior=psa" and ngspice is sensitive to placement of "X"

[ngspice_digital.zip](https://github.com/ra3xdh/qucs_s/files/14576878/ngspice_digital.zip)

**Reply by ra3xdh** [reply]
> Qucs-S is not writing the .spiceinit "ngbehavior=psa"

Qucs-S creates .spiceinit and writes its contents. I doesn't analyze the contents of the `spiceinit` block and transfers it to the spicenint file as is. Probably you have somewhere global spiceinit that conflicts with local spiceinit. There is no bug in Qucs-S.

**Reply by tomhajjar** [reply]
I got rid of my global .spiceinit weeks ago. I know Qucs-S writes the .spiceinit to spice4qucs.

Can you see why only 1 of the netlists I sent works? For the files to run under DuSpice, I have to place .spiceinit in the ngspice-42 bin folder. I think there is an issue with ngspice.

**Comment by ra3xdh** [comment]
>  Is there a way to get them in the "sch" format so I can make libraries?

There is no way to export symbols of system devices. They are hardcoded in C++.

**Comment by tomhajjar** [comment]
Per your recommendation I redid my subcircuits using "Spice netlist" using individual 74HCxx.cir files copy and pasted from MicroCap dig000.lib and things seem to be working.

![2024-03-12_161235](https://github.com/ra3xdh/qucs_s/assets/102838604/264345c7-dd0c-4751-a89c-9e8384284964)

**Comment by tomhajjar** [comment]
I'm not having any luck getting the logic gates from the MicroCap "dig000.lib" to work in more complicated mixed-signal circuits like this Trainline Driver. 

To get the circuit to run I had to use "74HCng_gates.lib". These models use adc_bridge and dac_bridge. No 74HC74 so I had to simplify the circuit. Waiting for Holger.

Also the VOM1271T isn't fully turning on the IRFP250. Turns out I forgot the model doesn't work...

![2024-03-12_184703](https://github.com/ra3xdh/qucs_s/assets/102838604/6dbaa6d8-9ff3-4651-8f1f-87b8d50c8f4f)
![2024-03-12_213233](https://github.com/ra3xdh/qucs_s/assets/102838604/3deea2af-7452-4390-b5e6-87f5a5052118)
![2024-03-12_213149](https://github.com/ra3xdh/qucs_s/assets/102838604/3fa0bcb8-34f1-41ae-965f-3753d0a6d6d2)

**Comment by tomhajjar** [comment]
Holger is looking into the issues I am having. There is a bug causing issues with MicroCap "dig000.lib" when "reset" is in the netlist but it doesn't always cause issues. I can make netlists using "dig000.lib" that work and ones that don't work. To make "bad" ones work I have to delete all "reset" statements.

So far I'm not having issues with devices using adc_bridge and dac_bridge.

I updated the project above with the latest.

https://sourceforge.net/p/ngspice/discussion/120973/thread/e6442c59ce/?page=1

**Comment by ra3xdh** [comment]
There is something wrong with digital devices in Ngspice. I suspect the logic output resistance is simulated wrong. The following schematic doesn't converge, but the RC-oscillator and frequency divider work separately.
![image](https://github.com/ra3xdh/qucs_s/assets/4920080/dc3ae5d7-a93a-42f6-a98f-b21daddd9ffb)

**Comment by tomhajjar** [comment]
I altered your schematic to use async counter. Plots look "nicer".

As Holger said there is probably a bug in ngspice that cause the "reset" in the netlist to halt the digital simulation under certain circumstances. He also said Op doesn't work for "digital". I assume he means for "dig000.lib" and not "74HCng_gates.lib" which uses XSPICE.

I made a netlist in the ngspice folder "74HC00-74_osc_div_tran.cir" that looks to be working. Everything that "bothers" ngspice or deletes DuSpice screen plotting is commented out. 

Is there a reason Timing Diagrams don't work for these plots? I was forced to use Nutmeg equations to space the traces out.

![2024-03-15_091439](https://github.com/ra3xdh/qucs_s/assets/102838604/b408898f-3a1b-4d5e-b1a4-0aadc1ecb2cc)

![2024-03-15_091737](https://github.com/ra3xdh/qucs_s/assets/102838604/32c8147e-30cd-411b-b3ff-70badc2e87a6)

![2024-03-15_134245](https://github.com/ra3xdh/qucs_s/assets/102838604/c71f509e-e2c8-42d2-bd8e-ee80d9a16abc)

![2024-03-15_135157](https://github.com/ra3xdh/qucs_s/assets/102838604/a81d39cf-8770-4973-b62b-3d7b8750f91f)

**Comment by tomhajjar** [comment]
Updated Trainline_Driver_prj on 10/30/24

Holger uploaded a fix. He also made a new Windows nightly. More testing to do...

![2024-03-15_171912](https://github.com/ra3xdh/qucs_s/assets/102838604/6832eae9-98ae-4ab4-9696-79c01f7903ac)
![2024-03-15_225422](https://github.com/ra3xdh/qucs_s/assets/102838604/afb2c4d0-267d-4894-8f1b-f20a982b4bbe)
![2024-03-15_225408](https://github.com/ra3xdh/qucs_s/assets/102838604/67b414cd-177a-43d0-b677-2b23fad4ee9b)

[Trainline_Driver_prj.zip](https://github.com/user-attachments/files/17578355/Trainline_Driver_prj.zip)

**Comment by tomhajjar** [comment]
Updated Digital_Analog_Mixed_prj.zip 10/30/24


It's hit or miss finding TI behavioral/transient Spice models for specific digital devices. I prefer these for mixed signal analysis.

Having issues with TI 74LCV74 and 74HCS74 Spice models. Data is different depending on pulse widths as detailed in link below.
https://sourceforge.net/p/ngspice/discussion/120973/thread/e6442c59ce/?limit=25&page=2#37b8

![2024-03-20_183323](https://github.com/ra3xdh/qucs_s/assets/102838604/6430625e-f2c5-4656-b312-95360de0adb1)

[Digital_Analog_Mixed_prj.zip](https://github.com/user-attachments/files/17578305/Digital_Analog_Mixed_prj.zip)
